Support Arbitrary Number of Block Arguments

daumayr opened the pull request:
Expand Blocks so an arbitrary number of arguments can be accepted: `aBlock valueWithArguments: arg` where `arg` is an `Array` containing the arguments.

- change semantics for blocks, so that they throw an error when the number of arguments does not match
- added `#cull*` methods to allow blocks to ignore arguments
- remove the `BlockN` classes, which originated from SOM


This PR solves #94 and #216.

daumayr committed:
Added support for arbitary number of block arguments.

- added tests
- make blocks more strict, incorrect argument number will result in
  error

Co-authored-by: Stefan Marr <git@stefan-marr.de>

smarr commented on core-lib/Kernel.ns:
```diff
@@ -630,18 +630,21 @@ class Kernel vmMirror: vmMirror = Object <: Value (
     public value: arg1 with: arg2 = ( ^ vmMirror blockValue: self with: arg1 with: arg2 )
   )
 
+  protected class BlockArbitrary = Block ()
```
Do we at this point still need the other block classes?
Couldn't we just do away with Block1-3?

smarr replied:
@richard-roberts any opinions/requirements?
How does Grace handle block activations when the number of arguments doesn't match?

richard-roberts replied:
We just throw an error; from the Grace spec: "Requesting the apply(...) method evaluates the block; it is an error to provide the wrong number of arguments."

smarr commented on core-lib/Kernel.ns:
```diff
@@ -64,7 +64,7 @@ class Kernel vmMirror: vmMirror = Object <: Value (
     public notNil = ( ^ true )
     public ifNil: aBlock = ( ^ self )
     public ifNotNil: aBlock = ( ^ aBlock value )
-    public ifNil: noGoBlock ifNotNil: goBlock = ( ^ goBlock value: self )
```
This was actually correct. But `ifNotNil:` seems wrong.
The reason is that we want to capture the value and have it in the block argument:

```Smalltalk
obj ifNotNil: [:it | it isGuaranteedToExist ].
(* or also *)
obj ifNil: [ #foo ] ifNotNil: [:it <SomeThing> | it baz ].
```

This is used for instance in MiniTest.

smarr replied:
More digging gives both uses, with, and without `:it`.
Think we need `cull:` then to give a missing value as `nil`.

smarr replied:
Something to consider: http://forum.world.st/cull-protocol-td1560122.html

daumayr replied:
What is your preference for implementing this: Kernel, Primitive, or both?

smarr replied:
First in Kernel, and if it turns out to be a performance concern, we can optimize it?

daumayr replied:
Sure, we need a numArgs primitive for that anyway, do we want to expose that for other use-cases or just culling?

smarr replied:
Good question. We do not expose it in Symbol either, and I'd consider it reflective information.
So strictly speaking it should be only exposed via mirrors.
Think the best would be to simply use a `vmMirror` primitive directly in the `cull*` implementations.

smarr committed:
Fix code style in LanguageTests.ns

- prefer explicit import over use via Kernel
- spaces in blocks
- no spaces between element and array separator
- use #should:signal: or #shouldnt:signal: for exception testing

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Removing BlockN classes

- update README to reflect lifted restriction

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Add special case for empty array literals (fixes #216)

This prevents issues with the empty expression array in the specialization logic.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added primitive to access the number of arguments of a block

This is needed for the implementation of #cull:*

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Added #cull: allowing blocks to ignore arguments

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr committed:
Update test data after kernel changes

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
I did a few more cleanups, updated the PR description, and added `#cull*` support.
